File: 125_algorithms/_exercises/templates/_algorithms_challenges/leetcode/leetCode/BinarySearch/74_Search2DMatrix.py
#! /usr/bin/env python
# -*- coding: utf-8 -*-


c.. Solution o..
    # Don't treat it as a 2D matrix, just treat it as a sorted list
    ___ searchMatrix  matrix, target
        __ n.. matrix:
            r_ False

        # Classic binary search: O(logmn)
        m_rows, n_cols = l..(matrix), l..(matrix[0])
        left, right = 0, m_rows * n_cols - 1

        _____ left <= right:
            mid = (left+right) / 2
            num = matrix[mid / n_cols][mid % n_cols]
            __ num > target:
                right = mid - 1
            ____ num < target:
                left = mid + 1
            ____
                r_ True

        r_ False

"""
[[]]
0
[[1]]
0
[[1, 3, 5, 7], [10, 11, 16, 20], [23, 30, 34, 50]]
34
[[1, 3, 5], [10, 11, 16], [23, 30, 34]]
46
"""

File: 125_algorithms/_exercises/templates/_algorithms_challenges/leetcode/leetCode/BinarySearch/74_Search2DMatrix.cpp
/*
 * @Author: xuezaigds@gmail.com
 * @Last Modified time: 2016-04-13 11:17:48
 */

c.. Solution {
public:
    // Don't treat it as a 2D matrix, just treat it as a sorted list
    bool searchMatrix(vector<vector<int>>& matrix, int target) {
        __(matrix.empty()){
            r_ false;
        }
        int m_rows = matrix.size(), n_cols = matrix[0].size();
        int left=0, right = m_rows*n_cols-1;
        _____(left <= right){
            int mid = left + (right-left) / 2;
            int num = matrix[mid/n_cols][mid%n_cols];
            __(num < target){
                left = mid + 1;
            }
            else __(num > target){
                right = mid - 1;
            }
            else{
                r_ true;
            }
        }
        r_ false;
    }
};

/*
[[]]
0
[[1]]
0
[[1, 3, 5, 7], [10, 11, 16, 20], [23, 30, 34, 50]]
34
[[1, 3, 5], [10, 11, 16], [23, 30, 34]]
46
*/
